### PR TITLE
New RGBColor Instant Answer

### PR DESCRIPTION
**Pull request by GuiltyDolphin**
This PR is for a new Instant Answer: `RGBColor`.

The IA will provide an Instant-Answer interface to the Color::RGB::Util Perl package, and link to the Color Picker for more advanced usage.

### To-Do

- [x] Mixing colors
- [x] Random colors
- [x] Reversing colors
- [x] Tinting
- [X] Pretty output (probably)
- [x] Linking to ColorPicker

### Future Plans

- When #3140 is merged, use it to support multiple colors in mixing and tinting.
- Add other fun stuff!

---

Instant Answer Page: https://duck.co/ia/view/rgb_color

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @GuiltyDolphin

### CODE REVIEW

**Comment by GuiltyDolphin** [comment]
@moollaza Phew! I think there's enough here to deploy to Beta; couple of deps though.

**Comment by moollaza** [comment]
Deployed!

**Comment by GuiltyDolphin** [comment]
@moollaza Awesome, thanks!

By the way, I've currently got the IA to link to the color picker with the input and result colors; though the color picker currently ignores any input colors (I've provided it in HTML Hex form). There's an open issue for this though so hopefully it'll be fixed soon :)

**Comment by GuiltyDolphin** [comment]
@moollaza Righty, there's nowt else I want to add just yet, so I think this is ready for review.

Hopefully this'll cover a lot of color queries :)

**Comment by GuiltyDolphin** [comment]
@moollaza I'm going to try and fix up the design to be more like ColorCodes - I feel that is rather appropriate (and would provide a more consistent feel).

**Comment by GuiltyDolphin** [comment]
@moollaza @mintsoft I'm pretty happy with this for the time-being; couple of things I want to do in the future, but apart from that this should be good for review.

cc @tagawa @edgesince84 any thoughts on the design? It's essentially the same as for the ColorCodes IA bar some modifications (there's a couple of changes I'd like to make in the future - but I'd just mess it up now :wink:)

**Comment by moollaza** [comment]
@GuiltyDolphin I'm seeing the wrapping date boxes, but only at a certain breaking point in the CSS

To prevent the `input`s from wrapping, you can wrap them in `<div class="one-line">` -- I just tested by modifying the DOM and it works 👍

**Comment by moollaza** [comment]
@GuiltyDolphin BTW the updates are now deployed to BETA

**Comment by GuiltyDolphin** [comment]
@moollaza Sorry? Date boxes?

Was that comment meant for DateMath?

**Comment by emanresusername** [comment]
not sure what the etiquette is here 😅 . . . but just saw this on beta and 
gif-keyboard-57088175041256600

**Comment by moollaza** [comment]
> Was that comment meant for DateMath?

Whoops! Yes it was. Must have got my tabs confused, sorry!

**Comment by moollaza** [comment]
@GuiltyDolphin https://beta.duckduckgo.com/?q=opposite+color+of+cyan&ia=answer looks a little odd.

You should be able to remove the set height and width of `column2` and also `.desc_column`

File: share/goodie/rgb_color/rgb_color.css [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    margin-left: -0.7em;
+}
+
+.zci--rgb_color  p.no_vspace {
+    margin: 0;
+    padding: 0;
+}
+
+.zci--rgb_color .column1
+{
```
Comment (moollaza) [inline share/goodie/rgb_color/rgb_color.css]: nit: should be on the line above

File: share/goodie/rgb_color/rgb_color.css [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    display: inline-block;
+    height: 125px;
+    padding-top: 7px;
+}
+
+.zci--rgb_color .column2 {
+	display: inline-block;
```
Comment (moollaza) [inline share/goodie/rgb_color/rgb_color.css]: nit: indentationi

File: share/goodie/rgb_color/rgb_color.css [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    margin-left: -2px;
+    margin-right: 4px;
+}
+
+.zci--rgb_color .desc_column {
+    display: inline-block;
+    width: 10em;
+}
+
+.zci--rgb_color .complementary {
```
Comment (moollaza) [inline share/goodie/rgb_color/rgb_color.css]: can be removed

File: lib/DDG/Goodie/RgbColor.pm [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    my $amt2 = 0.5;
+    if ($a1 // $a2) {
+        ($amt1, my $t1) = amount_type_from_text($a1);
+        ($amt2, my $t2) = amount_type_from_text($a2);
+        return unless $t1 eq $t2;
+        $pct = right_decimal_from_amount($amt1, $amt2, $t1) // return;
+        ($amt1, $amt2) = normalize_amounts_for_template($t1, $amt1, $amt2);
+    }
+    my %data = (
+        subtitle_prefix => 'Mix ',
```
Comment (moollaza) [inline lib/DDG/Goodie/RgbColor.pm]: I think `Mix #:` is better? (or even `Combine #:`)

File: lib/DDG/Goodie/RgbColor.pm [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+}
+
+sub reverse_color {
+    my %cap = @_;
+    my $c = normalize_color($cap{c});
+    my $color = normalize_color_for_template(reverse_rgb_color($c));
+    return (
+        data => {
+            subtitle_prefix => '(RGB) Opposite color of ',
+            input_colors    => [normalize_color_for_template($c)],
```
Comment (moollaza) [inline lib/DDG/Goodie/RgbColor.pm]: Output color names should be uppercase: e.g `0000FF`

File: lib/DDG/Goodie/RgbColor.pm [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+    return %result;
+}
+
+sub reverse_color {
+    my %cap = @_;
+    my $c = normalize_color($cap{c});
+    my $color = normalize_color_for_template(reverse_rgb_color($c));
+    return (
+        data => {
+            subtitle_prefix => '(RGB) Opposite color of ',
```
Comment (moollaza) [inline lib/DDG/Goodie/RgbColor.pm]: Perhaps `Opposite of #:`?

`subtitle_prefix` should end with a `:` -- we can remove the trailing space as well

File: lib/DDG/Goodie/RgbColor.pm (hunk 5 above)
Comment (moollaza) [inline lib/DDG/Goodie/RgbColor.pm]: You can use CSS for this if you put a new class in the template 👍

File: lib/DDG/Goodie/RgbColor.pm (hunk 4 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RgbColor.pm]: Hmm? Why the `#`?

I'm planning on sprucing the output up a little in the future (adding Hex signs + Names); are you happy to wait for that?

**Comment by GuiltyDolphin** [comment]
@moollaza Sorry about the CSS - I didn't do much with it after copying it over... Heh. :grin:

**Comment by GuiltyDolphin** [comment]
@moollaza Anything else?

**Comment by moollaza** [comment]
@GuiltyDolphin Thanks! This LGTM :+1:

I forgot to make a task for the internal dependencies though, so once that's complete this can be merged 👍

File: lib/DDG/Goodie/RgbColor.pm [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+my %colors = map { (lc $_->title =~ s/-/ /gr) => $_, lc $_->name => $_ } @dict_colors;
+my @color_descs = sort { length $b <=> length $a } keys %colors;
+
+my %hex_to_color = map { $_->hex => $_ } @dict_colors;
+my $color_name_re = '(?:' .
+    (join '|', map { quotemeta $_ } @color_descs) . ')';
+$color_name_re =~ s/\\ /[ -]?/g;
+
+my $scolor = 'colou?rs?';
+my $color_re = qr/(?:$color_name_re|#?\p{XDigit}{6})/;
```
Comment (moollaza) [inline lib/DDG/Goodie/RgbColor.pm]: @GuiltyDolphin can we allow `#[a-z]{3}` here?

I tried `random color between #eee and #fff` and didn't get any results.

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RgbColor.pm]: Done.

**Comment by GuiltyDolphin** [comment]
@moollaza How're the deps going? Ready to go?

**Comment by moollaza** [comment]
> Ready to go?

Not quite yet, waiting for ops team. It's a low priority for them.

**Comment by moollaza** [comment]
@GuiltyDolphin Thanks! This LGTM :+1: Deps were added this morning 👍

**Comment by daxtheduck** [comment]
@GuiltyDolphin this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Pull request merged**